plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExBudgetStream.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE BangPatterns #-}
+
+module PlutusCore.Evaluation.Machine.ExBudgetStream
+    ( ExBudgetStream(..)
+    , sumExBudgetStream
+    , zipCostStream
+    ) where
+
+import PlutusCore.Evaluation.Machine.CostStream
+import PlutusCore.Evaluation.Machine.ExBudget
+import PlutusCore.Evaluation.Machine.ExMemory
+
+import Data.Coerce
+
+-- | A lazy stream of 'ExBudget's. Basically @NonEmpty ExBudget@, except the elements are
+-- stored strictly.
+--
+-- The semantics of a stream are those of the 'fold' of its elements. I.e. a stream that is a
+-- reordered version of another stream is considered equal to that stream.
+--
+-- An 'ExBudgetStream' is what one gets by zipping two 'CostStream's (one for CPU, one for memory),
```

Comment:
I think `zipWith` counts as zipping? Its Haddock says

> [zipWith](https://hackage.haskell.org/package/base-4.18.0.0/docs/GHC-OldList.html#v:zipWith) generalises [zip](https://hackage.haskell.org/package/base-4.18.0.0/docs/GHC-OldList.html#v:zip) by zipping with the function given as the first argument, instead of a tupling function.